src/engineio/transports.rs

```diff
@@ -0,0 +1,129 @@
+    tls_config: Arc<Mutex<Option<TlsConnector>>>,
+    transport_type: Arc<RwLock<TransportTypes>>,
+}
+
+impl Transports {
+    pub fn new(tls_config: Option<TlsConnector>, opening_headers: Option<HeaderMap>) -> Self {
+        Transports {
+            websocket_secure: Arc::new(Mutex::new(None)),
+            websocket: Arc::new(Mutex::new(None)),
+            polling: Arc::new(Mutex::new(Some(PollingTransport::new(
+                tls_config.clone(),
+                opening_headers,
+            )))),
+            tls_config: Arc::new(Mutex::new(tls_config)),
+            transport_type: Arc::new(RwLock::new(TransportTypes::Polling)),
+        }
+    }
+
+    pub(super) fn upgrade_websocket(&mut self, address: String) -> Result<()> {
+        if self.websocket_secure.lock()?.is_none() {
+            *self.websocket.lock()? = Some(WebsocketTransport::new(address));
+            self.websocket.lock()?.as_ref().unwrap().probe()?;
+            *self.transport_type.write()? = TransportTypes::Websocket;
+            Ok(())
+        } else {
+            Err(Error::TransportExists())
+        }
+    }
+
+    pub(super) fn upgrade_websocket_secure(&mut self, address: String) -> Result<()> {
+        if self.websocket_secure.lock()?.is_none() {
+            *self.websocket_secure.lock()? = Some(WebsocketSecureTransport::new(
+                address,
+                self.tls_config.lock()?.clone(),
+            ));
+            self.websocket_secure.lock()?.as_ref().unwrap().probe()?;
+            *self.transport_type.write()? = TransportTypes::WebsocketSecure;
+            Ok(())
+        } else {
+            Err(Error::TransportExists())
+        }
+    }
+}
+
+impl Transport for Transports {
+    fn emit(&self, address: String, data: Bytes, is_binary_att: bool) -> Result<()> {
+        match &*self.transport_type.read()? {
+            TransportTypes::Websocket => {
+                self.websocket
+                    .lock()?
+                    .as_ref()
+                    .unwrap()
+                    .emit(address, data, is_binary_att)
+            }
+            TransportTypes::WebsocketSecure => self
+                .websocket_secure
+                .lock()?
+                .as_ref()
+                .unwrap()
+                .emit(address, data, is_binary_att),
+            TransportTypes::Polling => {
+                self.polling
+                    .lock()?
+                    .as_ref()
+                    .unwrap()
+                    .emit(address, data, is_binary_att)
+            }
+        }
+    }
+
+    fn poll(&mut self, address: String) -> Result<Bytes> {
+        match &*self.transport_type.read()? {
+            TransportTypes::Websocket => self
+                .websocket_secure
+                .lock()?
+                .as_mut()
+                .unwrap()
+                .poll(address),
+            TransportTypes::WebsocketSecure => {
+                self.websocket.lock()?.as_mut().unwrap().poll(address)
```

Comment:
Did you maybe mix up `Websocket` and `WebsocketSecure` here?

Reply:
Nice catch, fixed in https://github.com/1c3t3a/rust-socketio/pull/52/commits/b4a11b53a532fe8770c28308b4e79f2ae0d8d365